src/Compilers/Core/Rebuild/CompilationOptionsReader.cs

```diff
@@ -271,35 +284,75 @@ public ImmutableArray<MetadataReferenceInfo> GetMetadataReferences()
         {
-            var sourceFileCount = int.Parse(
-                GetMetadataCompilationOptions()
-                    .GetUniqueOption(CompilationOptionNames.SourceFileCount));
+            var syntaxTrees = ResolveSyntaxTrees();
+            var metadataReferences = ResolveMetadataReferences();
+            return (syntaxTrees, metadataReferences);
 
-            var builder = ImmutableArray.CreateBuilder<SourceFileInfo>(sourceFileCount);
-            foreach (var documentHandle in PdbReader.Documents.Take(sourceFileCount))
+            ImmutableArray<SyntaxTree> ResolveSyntaxTrees()
             {
-                var document = PdbReader.GetDocument(documentHandle);
-                var name = PdbReader.GetString(document.Name);
-
-                var hashAlgorithmGuid = PdbReader.GetGuid(document.HashAlgorithm);
-                var hashAlgorithm =
-                    hashAlgorithmGuid == HashAlgorithmSha1 ? SourceHashAlgorithm.Sha1
-                    : hashAlgorithmGuid == HashAlgorithmSha256 ? SourceHashAlgorithm.Sha256
-                    : SourceHashAlgorithm.None;
+                var sourceFileCount = GetSourceFileCount();
+                var builder = ImmutableArray.CreateBuilder<SyntaxTree>(sourceFileCount);
+                foreach (var tuple in GetSourceTextInfoCore())
+                {
+                    SourceText sourceText;
+                    if (ResolveEmbeddedSource(tuple.DocumentHandle, tuple.SourceTextInfo) is { } embeddedSourceTextInfo)
+                    {
+                        sourceText = embeddedSourceTextInfo.SourceText;
+                    }
+                    else
+                    {
+                        sourceText = resolver.ResolveSourceText(tuple.SourceTextInfo);
+                        if (!sourceText.GetChecksum().SequenceEqual(tuple.SourceTextInfo.Hash))
+                        {
+                            throw new InvalidOperationException();
+                        }
+                    }
 
-                var hash = PdbReader.GetBlobBytes(document.Hash);
-                var embeddedContent = ResolveEmbeddedSource(documentHandle, hashAlgorithm, encoding);
+                    var syntaxTree = createSyntaxTreeFunc(tuple.SourceTextInfo.OriginalSourceFilePath, sourceText);
+                    builder.Add(syntaxTree);
+                }
 
-                builder.Add(new SourceFileInfo(name, hashAlgorithm, hash, embeddedContent.embeddedText, embeddedContent.compressedHash));
+                return builder.MoveToImmutable();
             }
 
-            return builder.MoveToImmutable();
+            ImmutableArray<MetadataReference> ResolveMetadataReferences()
+            {
+                var builder = ImmutableArray.CreateBuilder<MetadataReference>();
+                foreach (var metadataReferenceInfo in GetMetadataReferenceInfo())
+                {
+                    var metadataReference = resolver.ResolveMetadataReference(metadataReferenceInfo);
```

Comment:
>var metadataReference = resolver.ResolveMetadataReference(metadataReferenceInfo) [](start = 20, length = 80)

Looking at the implementation of `CompilationRebuildArtifactResolver`, it looks like this `foreach` loop is O(N^2). Will we hit cases where there are enough metadata references to make that an issue?

Reply:
The `CompilationRebuildArtifactResolver` in particular I'm not worried about. That is a test only type. The underlying algorithm I'm interested in though. 

At the core the host has to provide matching artifacts for the underlying specification. Generally the core lookup routine is done by Hash and MVID (for source and reference respectively). I would expect a production resolver to have close to O(1) lookup for each artifact. 

Consider that `BuildValidator` does this for `SourceText` because it knows to lookup based on the file path first, then do the hash check. The lookup for references is not that fast because that's a harder problem to solve for the local build case. Even then it creates MVID based caches for future lookup so over the course of validating Roslyn the O(N) lookup per reference should be amortized away.

Reply:
Ah, I'd overlooked the fact that `CompilationRebuildArtifactResolver` was test code.

---
In reply to: [610178574](https://github.com/dotnet/roslyn/pull/52511#discussion_r610178574) [](ancestors = 610178574)